[CI/Build] Allow hermetic builds

fabiendupont opened the pull request:
The current Dockerfile assumes that many build artifacts are available from public repositories and downloads them from these repositories, making it more difficult for downstream distributions to perform hermetic builds of vLLM container images.

This pull request introduces changes that should be backward compatible, while improving the situation for hermetic builds. Below is the list of changes:

- Build argument for the base images. Currently both the build and final images are based on NVIDIA CUDA devel images, but with different Ubuntu versions.
- Support for a mirror of the Deadsnakes PPA. This also requires installing the GPG key manually, as it's deprecated in add-apt-repository.
- Installation of pip from a pyz archive. This is an alternative to get-pip.py that allows using a private copy of the pip.pyz file.
- Alternative Python indexes. The {PIP,UV}_INDEX_URL and {PIP,UV}_EXTRA_INDEX_URL variables are native. Reusing the same naming convention for CUDA and CUDA nightly indexes.
- Custom URL and endpoint for sccache. The URL is for installation from an internal repository. The endpoint allows storing the cache in a non-AWS S3 backend.
- Custom FlashInfer installation. The change allows using package name+version, a Git URL or even a local .whl file.
- Add pip/uv authentication with keyring. The pip/uv command supports a {PIP,UV}_KEYRING_PROVIDER and it is set to disabled by default.

Co-authored-by: Elias Levy <eliaslevy@google.com>

<!--- pyml disable-next-line no-emphasis-as-heading -->

fabiendupont committed:
Parameterize base images

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

fabiendupont committed:
Allow Deadsnakes repository mirror

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

fabiendupont committed:
Use a build arg for get-pip.py URL

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

fabiendupont committed:
Allow using custom Python indexes

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

fabiendupont committed:
Allow custom sscache source and endpoint

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

fabiendupont committed:
Allow custom FlashInfer version and location

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

fabiendupont committed:
Add pip authentication with keyring

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

tlrmchlsmth commented on docker/Dockerfile:
```diff
@@ -6,30 +6,61 @@
 # docs/source/assets/contributing/dockerfile-stages-dependency.png
 
 ARG CUDA_VERSION=12.8.1
+ARG BUILD_BASE_IMAGE=nvidia/cuda:${CUDA_VERSION}-devel-ubuntu20.04
+ARG FINAL_BASE_IMAGE=nvidia/cuda:${CUDA_VERSION}-devel-ubuntu22.04
```
See my comment here -- I'm concerned about upgrading to Ubuntu 22.04 due to glibc issues

https://github.com/vllm-project/vllm/pull/18092#issuecomment-2877805116

tlrmchlsmth reviewed (commented):
Is there any way to unit test this?

tlrmchlsmth reviewed (approved):
LGTM overall and makes sense. I am a little concerned about fragility -- it isn't clear from looking at the Dockerfile _why_ these variables are factored out into arguments, so I could easily see someone breaking the hermetic builds. I suggest adding some inline comments.

@khluu could you please take a look as well?

fabiendupont commented:
@tlrmchlsmth, that's a fair point. I have added comments for the build arguments.

Regarding unit testing, I'm not sure, unless we want to setup mirrors for the various artifacts. I have mostly focused on non-regression for the existing build system.

tlrmchlsmth committed:
Fix pre-commit

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

DarkLight1337 commented:
`test_rejection_sampler.py` is not failing on main, so I think it's related to this PR

DarkLight1337 commented on docker/Dockerfile:
```diff
@@ -252,6 +401,8 @@ RUN --mount=type=bind,from=build,src=/workspace/dist,target=/vllm-workspace/dist
 # $ # upload the wheel to a public location, e.g. https://wheels.vllm.ai/flashinfer/524304395bd1d8cd7d07db083859523fcaa246a4/flashinfer_python-0.2.1.post1+cu124torch2.5-cp38-abi3-linux_x86_64.whl
 
+# Allow specifying a version, Git revision or local .whl file
+ARG FLASHINFER_WHEEL="git+https://github.com/flashinfer-ai/flashinfer@v0.2.2.post1"
```
Probably you should use the commit-specific wheel for flashinfer

fabiendupont replied:
Good point. I changed it to the commit from main.

fabiendupont committed:
Factor out common args and value

The default values for the build args should be specified only
once to reduce risk of desynchronization. This commit also moves
the comments to the top of the Dockerfile to avoid text
duplication.

Co-authored-by: Elias Levy <eliaslevy@google.com>
Signed-off-by: Fabien Dupont <fabiendupont@pm.me>

fabiendupont committed:
Add build args for UV

UV is using its own environment variables to set the package
indexes. This commit adds the UV_INDEX_URL, UV_EXTRA_INDEX_URL
and UV_KEYRING_PROVIDER. The defaults are copied from the PIP
build args for convenience.

Co-authored-by: Elias Levy <eliaslevy@google.com>
Signed-off-by: Fabien Dupont <fabiendupont@pm.me>

hmellor committed:
Fix dockerfile graph image

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

fabiendupont commented:
@ProExpertProg I am planning to add them later. We have internal patches to avoid downloading during the build. Do you prefer that I add them here? That's doable.

ProExpertProg commented:
> @ProExpertProg, I checked and we're actually using the following environment variables to control the CMake FetchContent dependencies: VLLM_CUTLASS_SRC_DIR, VLLM_FLASH_ATTN_SRC_DIR and FLASH_MLA_SRC_DIR. When they are set, CMake doesn't download the source at build time.
> 
> The challenge is to conditionally copy the content from local or remote locations at build time, as the Dockerfile syntax doesn't have conditionals.
> 
> All that to say that I'd rather address that in a separate PR.

Sounds good to me! Thanks for checking.

youkaichao reviewed (commented):
hmmm this is sooo intrusive and makes future maintenance and update quite difficult.

I think you can just build the docker image with internet access, and port the image directly for your use case?

fabiendupont committed:
Some robustness fixes

- Remove extra white spaces after '\' on end of line. This shouldn't be
  an issue with the current command, but could make troubleshooting more
  difficult.

- Use PYTHON_VERSION variable when copying to Python paths.

Signed-off-by: Fabien Dupont <fdupont@redhat.com>

smarterclayton commented:
@youkaichao 

I think the point of the change is to improve the security and reproducibility story for all users of vllm.  The current status quo results in a build process that is highly vulnerable to supply chain attacks which could leak sensitive models when vLLM is used.  Since those users will want to improve their resilience to such attacks by controlling dependencies more strongly (including the items that Fabian is changing here), the steps in this PR bring some of that control back to this project, vs being completely dependent on our upstreams for good dependency hygiene.

I agree this increases the complexity of the core vLLM image build process - however it is fairly normal for OSS projects like vLLM to strongly control how they accept dependencies, and to work to ensure they are consumable by downstream OSS / closed efforts (such as Linux distros).  vLLM is at much higher risk than many other projects because of the breadth of the dependency types and structure it is pulling in - in my experience parameterization of the Dockerfile represents a decent first step, even if there are further steps that would mitigate that complexity that Fabian is not yet taking.

fabiendupont commented:
@simon-mo, I am not sure it's because of the parameterization. The CUDA 12.1 and 11.8 versions don't support the 10.0 capabilities, which were added in 12.4. So, maybe we should drop CUDA <= 12.4 on the main branch.

aarnphm commented on docker/Dockerfile:
```diff
@@ -6,30 +6,106 @@
+# Example:
+# docker build --build-arg BUILD_BASE_IMAGE=registry.acme.org/mirror/nvidia/cuda:${CUDA_VERSION}-devel-ubuntu20.04
+ARG BUILD_BASE_IMAGE=nvidia/cuda:${CUDA_VERSION}-devel-ubuntu20.04
+ARG FINAL_BASE_IMAGE=nvidia/cuda:${CUDA_VERSION}-devel-ubuntu22.04
```
Can we make the base image for vllm-openai to be just the runtime cuda image, whereas we can maybe keep the devel image for test and CI.

fabiendupont replied:
This PR is not meant to change the behaviour of the current build, only allow using internal URL for hermetic builds. Not saying it's a bad idea, only not the purpose, so it could be a follow-up PR.

fabiendupont commented on docker/Dockerfile:
```diff
@@ -6,30 +6,106 @@
+# PIP, but they can be overridden.
+ARG PIP_INDEX_URL
+ARG PIP_EXTRA_INDEX_URL
+ARG UV_INDEX_URL=${PIP_INDEX_URL}
+ARG UV_EXTRA_INDEX_URL=${PIP_EXTRA_INDEX_URL}
```
Do we want to register the env inside the container. The index URLs are used to install packages during the image build, but at runtime the user doesn't install more Python packages, right?

The current Dockerfile doesn't set them, so default values from PIP are used, but the extra indexes were only passed on the `pip install` command lines, not persisting in the final image env.

aarnphm replied:
This makes sense. I would imagine for end consumer of the image it would depend on them controlling this
